### PR TITLE
Longest contribution streak bug

### LINKED ISSUE #9: Longest contribution streak bug
If you longest contribution streak is your current one you should not have an endedAt date.

### LINKED ISSUE #36: Tests: GetLongestContributionStreakByUsername
Tests: GetLongestContributionStreakByUsername

https://github.com/adavila0703/devy/blob/6fe79f88ef4d479d6ef45f63d655a35d371ec492/internal/github/contributions_test.go#L284

```golang

package github

import (
	"context"
	"fmt"
	"testing"
	"time"

	"github.com/shurcooL/githubv4"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
)

func TestGetContributionsByUsername(t *testing.T) {
	assert := assert.New(t)
	githubClient := &MockGithubClient{}
	githubService := NewGithubService(githubClient)

	ctx := context.Background()
	from := time.Date(2019, 1, 1, 0, 0, 0, 0, time.UTC)
	to := time.Date(2020, 1, 1, 0, 0, 0, 0, time.UTC)

	options := GetContributionsByUsernameOptions{
		Username: "devy",
		From:     from,
		To:       to,
	}

	githubClient.On(
		"Query",
		ctx,
		mock.AnythingOfType("*github.contributionsQuery"),
		mock.MatchedBy(func(params map[string]interface{}) bool {
			assert.Equal(githubv4.String(options.Username), params["username"])
			assert.Equal(githubv4.DateTime{Time: from}, params["from"])
			assert.Equal(githubv4.DateTime{Time: to}, params["to"])
			return true
		}),
	).Return(nil).Run(func(args mock.Arguments) {
		a := args.Get(1).(*contributionsQuery)
		(*a) = contributionsQuery{
			User: user{
				ContributionsCollection: contributionsCollection{
					ContributionCalendar: contributionCalendar{
						TotalContributions: 100,
						Weeks: []week{
							{
								[]contributionDays{
									{
										ContributionCount: 5,
										Weekday:           1,
										Date:              "2019-01-01",
									},
								},
							},
							{
								[]contributionDays{
									{
										ContributionCount: 10,
										Weekday:           0,
										Date:              "2019-01-02",
									},
								},
							},
						},
					},
				},
			},
		}
	}).Once()

	resp, err := githubService.GetContributionsByUsername(ctx, options)

	assert.NoError(err)

	assert.Equal(10, resp.Days[0].ContributionCount)
	assert.Equal(0, resp.Days[0].Weekday)
	assert.Equal(time.Date(2019, 01, 02, 0, 0, 0, 0, time.UTC), resp.Days[0].Date)

	assert.Equal(5, resp.Days[1].ContributionCount)
	assert.Equal(1, resp.Days[1].Weekday)
	assert.Equal(time.Date(2019, 01, 01, 0, 0, 0, 0, time.UTC), resp.Days[1].Date)

	githubClient.AssertExpectations(t)
}

func TestGetContributionsByUsername_MultiYear(t *testing.T) {
	assert := assert.New(t)
	githubClient := &MockGithubClient{}
	githubService := NewGithubService(githubClient)

	ctx := context.Background()
	from := time.Date(2020, 12, 6, 0, 0, 0, 0, time.UTC)
	to := time.Date(2021, 12, 7, 0, 0, 0, 0, time.UTC)

	options := GetContributionsByUsernameOptions{
		Username: "devy",
		From:     from,
		To:       to,
	}

	fmt.Println("check here", from)

	githubClient.On(
		"Query",
		ctx,
		mock.AnythingOfType("*github.contributionsQuery"),
		mock.MatchedBy(func(params map[string]interface{}) bool {
			return githubv4.String(options.Username) == params["username"] &&
				githubv4.DateTime{Time: time.Date(2020, 12, 7, 0, 0, 0, 0, time.UTC)} == params["from"] &&
				githubv4.DateTime{Time: time.Date(2021, 12, 7, 0, 0, 0, 0, time.UTC)} == params["to"]
		}),
	).Return(nil).Run(func(args mock.Arguments) {
		a := args.Get(1).(*contributionsQuery)
		(*a) = contributionsQuery{
			User: user{
				ContributionsCollection: contributionsCollection{
					ContributionCalendar: contributionCalendar{
						TotalContributions: 100,
						Weeks: []week{
							{
								[]contributionDays{
									{
										ContributionCount: 5,
										Weekday:           1,
										Date:              "2019-01-01",
									},
								},
							},
						},
					},
				},
			},
		}
	}).Once()

	githubClient.On(
		"Query",
		ctx,
		mock.AnythingOfType("*github.contributionsQuery"),
		mock.MatchedBy(func(params map[string]interface{}) bool {
			return githubv4.String(options.Username) == params["username"] &&
				githubv4.DateTime{Time: time.Date(2020, 12, 6, 0, 0, 0, 0, time.UTC)} == params["from"] &&
				githubv4.DateTime{Time: time.Date(2020, 12, 6, 0, 0, 0, 0, time.UTC)} == params["to"]
		}),
	).Return(nil).Run(func(args mock.Arguments) {
		a := args.Get(1).(*contributionsQuery)
		(*a) = contributionsQuery{
			User: user{
				ContributionsCollection: contributionsCollection{
					ContributionCalendar: contributionCalendar{
						TotalContributions: 100,
						Weeks: []week{
							{
								[]contributionDays{
									{
										ContributionCount: 5,
										Weekday:           1,
										Date:              "2019-01-01",
									},
								},
							},
						},
					},
				},
			},
		}
	}).Once()

	resp, err := githubService.GetContributionsByUsername(ctx, options)

	fmt.Println(resp.TotalContributions)

	assert.NoError(err)

	assert.Equal(5, resp.Days[0].ContributionCount)
	assert.Equal(1, resp.Days[0].Weekday)
	assert.Equal(time.Date(2019, 01, 01, 0, 0, 0, 0, time.UTC), resp.Days[0].Date)
	assert.Equal(200, resp.TotalContributions)

	githubClient.AssertExpectations(t)
}

func TestGetContributionsByUsername_DatesZeroValue(t *testing.T) {
	assert := assert.New(t)
	githubClient := &MockGithubClient{}
	githubService := NewGithubService(githubClient)

	ctx := context.Background()

	options := GetContributionsByUsernameOptions{
		Username: "devy",
	}

	githubClient.On(
		"Query",
		ctx,
		mock.AnythingOfType("*github.contributionsQuery"),
		mock.MatchedBy(func(params map[string]interface{}) bool {
			assert.WithinDuration(time.Now(), params["to"].(githubv4.DateTime).Time, time.Millisecond)
			assert.WithinDuration(time.Now().AddDate(-1, 0, 0), params["from"].(githubv4.DateTime).Time, time.Millisecond)
			return githubv4.String(options.Username) == params["username"]
		}),
	).Return(nil).Run(func(args mock.Arguments) {
		a := args.Get(1).(*contributionsQuery)
		(*a) = contributionsQuery{
			User: user{
				ContributionsCollection: contributionsCollection{
					ContributionCalendar: contributionCalendar{
						TotalContributions: 100,
						Weeks: []week{
							{
								[]contributionDays{
									{
										ContributionCount: 5,
										Weekday:           1,
										Date:              "2019-01-01",
									},
								},
							},
							{
								[]contributionDays{
									{
										ContributionCount: 10,
										Weekday:           0,
										Date:              "2019-01-02",
									},
								},
							},
						},
					},
				},
			},
		}
	}).Once()

	resp, err := githubService.GetContributionsByUsername(ctx, options)

	assert.NoError(err)

	assert.Equal(10, resp.Days[0].ContributionCount)
	assert.Equal(0, resp.Days[0].Weekday)
	assert.Equal(time.Date(2019, 01, 02, 0, 0, 0, 0, time.UTC), resp.Days[0].Date)

	assert.Equal(5, resp.Days[1].ContributionCount)
	assert.Equal(1, resp.Days[1].Weekday)
	assert.Equal(time.Date(2019, 01, 01, 0, 0, 0, 0, time.UTC), resp.Days[1].Date)

	githubClient.AssertExpectations(t)
}

// TODO Tests: error table test on GetContributionsByUsername
// labels: tests, good first issue
// Need to run a table test on GetContributionsByUsername to hit
// ErrMissingUsername and ErrToDateBeforeFromDate
func TestGetContributionsByUsername_Errors(t *testing.T) {

}

// TODO Tests: GetFirstContributionYearByUsername
// labels: tests, good first issue
func TestGetFirstContributionYearByUsername(t *testing.T) {

}

// TODO Tests: CurrentContributionStreak.String()
// labels: tests
func TestCurrentContributionStreak_String(t *testing.T) {

}

// TODO Tests: GetCurrentContributionStreakByUsername
// labels: tests
func TestGetCurrentContributionStreakByUsername(t *testing.T) {

}

// TODO Tests: LongestContributionStreak.String()
// labels: tests, good first issue
func TestLongestContributionStreak_String(t *testing.T) {

}

// TODO Tests: GetLongestContributionStreakByUsername
// labels: tests
func TestGetLongestContributionStreakByUsername(t *testing.T) {

}

// TODO Tests: TotalContribution.String()
// labels: tests, good first issue
func TestTotalContribution_String(t *testing.T) {

}

// TODO Tests: GetTotalContributionsByUsername
// labels: tests, good first issue
func TestGetTotalContributionsByUsername(t *testing.T) {

}

```

c363e90cb277c0c60672a5a149c41141afa55a9d

### PR DESCRIPTION
**Pull request by adavila0703**
- Fixed issue where if the longest streak is the current streak, there would be an end date.
- Added message to handle this new condition
- Added tests for GetLongestContributionStreakByUsername

Closes #9 
Closes #36

### CODE REVIEW

File: internal/github/contributions.go [hunk 1]
```diff
@@ -219,12 +225,19 @@ type LongestContributionStreak struct {
 }
 
 func (c LongestContributionStreak) String() string {
+	emptyTime := time.Time{}
+	if c.EndedAt == emptyTime {
```
Comment (DillonStreator) [inline internal/github/contributions.go:229]: Could use c.EndedAt.IsZero() instead hehe

File: internal/github/contributions.go [hunk 2]
```diff
@@ -276,6 +291,13 @@ func (g *GithubService) GetLongestContributionStreakByUsername(ctx context.Conte
 		streak++
 	}
 
+	now := time.Date(time.Now().Year(), time.Now().Month(), time.Now().Day(), 0, 0, 0, 0, time.UTC)
+	if longestContributionStreak.EndedAt == now {
+		longestContributionStreak.EndedAt = time.Time{}
+	}
+
+	fmt.Println(longestContributionStreak.EndedAt)
```
Comment (DillonStreator) [inline internal/github/contributions.go:299]: Could cleanup test printlns?

File: internal/github/contributions.go (hunk 1 above)
Comment (adavila0703) [reply internal/github/contributions.go:229]: Good catch haha

File: internal/github/contributions_test.go [hunk 3]
```diff
@@ -287,10 +319,252 @@ func TestLongestContributionStreak_String(t *testing.T) {
...
+
+	username := "devy"
+	year := time.Now()
+
+	githubClient.On(
+		"Query",
+		ctx,
+		mock.AnythingOfType("*github.contributionYears"),
+		mock.MatchedBy(func(params map[string]interface{}) bool {
+			assert.Equal(githubv4.String(username), params["username"])
```
Comment (DillonStreator) [inline internal/github/contributions_test.go:337]: Should return this assertion?

File: internal/github/contributions_test.go [hunk 4]
```diff
@@ -287,10 +319,252 @@ func TestLongestContributionStreak_String(t *testing.T) {
...
```
Comment (DillonStreator) [inline internal/github/contributions_test.go:431]: Return this assertion?

File: internal/github/contributions_test.go [hunk 5]
```diff
@@ -287,10 +319,252 @@ func TestLongestContributionStreak_String(t *testing.T) {
...
+	}).Once()
+
+	from := time.Date(year.Year(), 1, 1, 0, 0, 0, 0, time.UTC)
 
+	githubClient.On(
+		"Query",
+		ctx,
+		mock.AnythingOfType("*github.contributionsQuery"),
+		mock.MatchedBy(func(params map[string]interface{}) bool {
+			assert.WithinDuration(from, params["from"].(githubv4.DateTime).Time, time.Millisecond)
```
Comment (DillonStreator) [inline internal/github/contributions_test.go:454]: Maybe we can make a method which returns bool if two dates are within a duration. Could live in date package. Dig into what asser.WithinDuration is doing and replicate it. This way we can directly return the boolean expression in the matchedby which would be more correct.

Comment (adavila0703) [reply internal/github/contributions_test.go:454]: I agree, lets make an issue for this

File: internal/github/contributions.go [hunk 6]
```diff
@@ -219,12 +225,18 @@ type LongestContributionStreak struct {
...
+	}
 	return fmt.Sprintf("longest contribution streak: %d days started at: %s ended at: %s", c.Streak, c.StartedAt.Format("2006-01-02"), c.EndedAt.Format("2006-01-02"))
 }
 
 func (g *GithubService) GetLongestContributionStreakByUsername(ctx context.Context, username string) (*LongestContributionStreak, error) {
 	year, err := g.GetFirstContributionYearByUsername(ctx, username)
+
+	fmt.Println(year)
 	if err != nil {
+		fmt.Println("error")
```
Comment (DillonStreator) [inline internal/github/contributions.go:239]: Cleanup fmt.Printlns

File: internal/github/contributions.go [hunk 7]
```diff
@@ -219,12 +225,18 @@ type LongestContributionStreak struct {
...
+	if c.EndedAt.IsZero() {
+		return fmt.Sprintf("longest and current contribution streak: %d days started at: %s", c.Streak, c.StartedAt.Format("2006-01-02"))
+	}
 	return fmt.Sprintf("longest contribution streak: %d days started at: %s ended at: %s", c.Streak, c.StartedAt.Format("2006-01-02"), c.EndedAt.Format("2006-01-02"))
 }
 
 func (g *GithubService) GetLongestContributionStreakByUsername(ctx context.Context, username string) (*LongestContributionStreak, error) {
 	year, err := g.GetFirstContributionYearByUsername(ctx, username)
+
+	fmt.Println(year)
```
Comment (DillonStreator) [inline internal/github/contributions.go:237]: Cleanup fmt.Printlns

**Review by DillonStreator — APPROVED** [review]

**Pull request merged**